node/core/dispute-coordinator/src/participation/queues/mod.rs

```diff
@@ -159,94 +144,76 @@ impl Queues {
 		priority: ParticipationPriority,
 		req: ParticipationRequest,
 	) -> Result<()> {
-		let comparator = match priority {
-			ParticipationPriority::BestEffort => None,
-			ParticipationPriority::Priority =>
-				CandidateComparator::new(sender, &req.candidate_receipt).await?,
-		};
-		self.queue_with_comparator(comparator, req)?;
+		let comparator = CandidateComparator::new(sender, &req.candidate_receipt).await?;
+
+		self.queue_with_comparator(comparator, priority, req)?;
 		Ok(())
 	}
 
-	/// Get the next best request for dispute participation
-	///
-	/// if any.  Priority queue is always considered first, then the best effort queue based on
-	/// `added_count`.
+	/// Get the next best request for dispute participation if any.
+	/// First the priority queue is considered and then the best effort one.
 	pub fn dequeue(&mut self) -> Option<ParticipationRequest> {
 		if let Some(req) = self.pop_priority() {
 			// In case a candidate became best effort over time, we might have it also queued in
 			// the best effort queue - get rid of any such entry:
-			self.best_effort.remove(req.candidate_hash());
-			return Some(req)
+			self.best_effort.remove(&req.0);
+			return Some(req.1)
 		}
-		self.pop_best_effort()
+		self.pop_best_effort().map(|d| d.1)
 	}
 
 	fn queue_with_comparator(
 		&mut self,
-		comparator: Option<CandidateComparator>,
+		comparator: CandidateComparator,
+		priority: ParticipationPriority,
 		req: ParticipationRequest,
 	) -> std::result::Result<(), QueueError> {
-		if let Some(comparator) = comparator {
+		if priority.is_priority() {
 			if self.priority.len() >= PRIORITY_QUEUE_SIZE {
 				return Err(QueueError::PriorityFull)
 			}
 			// Remove any best effort entry:
-			self.best_effort.remove(&req.candidate_hash);
+			self.best_effort.remove(&comparator);
 			self.priority.insert(comparator, req);
 		} else {
 			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
 				return Err(QueueError::BestEffortFull)
 			}
 			// Note: The request might have been added to priority in a previous call already, we
 			// take care of that case in `dequeue` (more efficient).
```

Comment:
This is actually no longer true. Since we are using a `CandidateComparator` for both queues now, we can refrain from putting a participation request in the best effort queue if already present in the priority queue efficiently and don't need to resort to handling that in `dequeue`.

Reply:
Is this comment related to the `remove` in `dequeue` only? On line 159: https://github.com/paritytech/polkadot/blob/ab790ff09e3b73a93fae2559b9694ae4f66e7e15/node/core/dispute-coordinator/src/participation/queues/mod.rs#L159

Reply:
Done.

Reply:
Ok, I think there is a misunderstanding. Here in this else we should check whether the participation is already queued in priority and if so, skip enqueuing in the best effort queue.

Reply:
It is the dual to the removal from the best effort above:

1. If it is already in best effort - remove it there if now enqueued in priority.
2. If it is already in priority, skip enqueuing in best-effort. 

How can this even happen? In case a participation request stays in the queue for a bit, dispute gets confirmed and finality kicks in - then we might queue again, but this time as best-effort instead of priority.

There is an argument that at this point it should not be priority anymore, but in any case we should only have it enqueued on one of the two queues. We might want to change 2 to:

2.  If it is already in priority - remove it there if now enqueued in best-effort.

So logic would be the same as for priority. It is a corner case and should not matter much in practice, so I don't mind either way.

Reply:
Ohh right. There even is a test for it.

Reply:
Should be fine now.